### ISSUE TITLE
error squiggle when using `%pip` in interactive window

### ISSUE THREAD

**Issue by amunger** [microsoft/pylance-release#2894]
enter `%pip install torch` into the IW input box

![image](https://user-images.githubusercontent.com/2019016/172238929-cff4e782-32fb-4fc7-ae5d-55a0a65e7c04.png)

**Comment by greazer** [comment]
We think this may be due to pylance not realizing that the IW input box is part of a "notebook". The "pylance lsp notebooks" setting was on.

**Comment by debonte** [comment]
When LSP notebooks are enabled, the `_isNotebookCell` check in `NotebookHandler.handleOpen` returns `false` and therefore we set `ipythonMode` to `false` for the interactive window document.

**Comment by rchiodo** [comment]
No `_isNotebookCell` is actually wrong or should be updated to also include the interactive cell scheme.

**Comment by debonte** [comment]
@rchiodo, are you planning to have the interactive window be treated as a notebook document from LSP's perspective at some point in the future?

I could fix the %magic issue, but that's not the only problem when the LSP notebooks experiment is enabled. Since the interactive window cells are passed as individual text documents, Pylance doesn't chain them together as we do for notebook cells. And therefore IntelliSense, hover, semantic coloring, go to def, etc don't work when working with symbols defined in earlier cells.

![image](https://user-images.githubusercontent.com/12820898/172963252-3e80e4d4-47e1-4bec-a47d-fee6327033f5.png)

**Comment by rchiodo** [comment]
Sounds like a bug in the LSP client if it isn't already behaving this way. 

With the old middleware solution, the interactive window was treated just like a notebook. I concatenated all of the cells (including the special one at the bottom) into the concatenated document. 

This would need to happen in the new way too.

**Comment by debonte** [comment]
Ok, I'll take a look at the LSP code.

Earlier I asked:

> Should we be getting a notebookDocument/didOpen instead of a textDocument/didOpen?

And your reply was "No..." so I thought you were saying this was expected at the moment.

**Comment by rchiodo** [comment]
I would expect that the interactive window will have to be treated special in the language client. It's internally a notebook and separate text document, so unless Dirk pieced them together, the language client probably doesn't know about it.

**Comment by debonte** [comment]
> so unless Dirk pieced them together, the language client probably doesn't know about it

Is the interactive window a Python-specific feature? I'm wondering if it is generic enough that Dirk would be willing to handle this in vscode-languageclient or if Pylance should special-case it.

For IW cells (not the input box), part of the problem in my screenshot above was that Pylance's `notebookSelector` only requested notifications about Jupyter notebooks. [Fixing that](https://github.com/debonte/pyrx/commit/2bf477fa1eee346a945ae0b12373d2520816a756) got hover and semantic highlighting working, but go to definition is still broken (opens the target cell in a new tab).

**Comment by dbaeumer** [comment]
The notebook support in LSP syncs notebooks and not individual cells. Piecing this together might be complicated for me since currently a cell text document is referenced by notebook cell. when syncing things.

It sounds like that the interactive windows is visually part of a notebook. Why can't it be a notebook cell as well with a special language id. The you could simply use a special notebook selector (https://github.com/microsoft/vscode-languageserver-node/blob/12900cfb006ce56360e3f64feaeff9c0224854a1/protocol/src/common/protocol.notebook.ts#L312) and everything should work.

**Comment by rchiodo** [comment]
> Why can't it be a notebook cell as well with a special language id. 

It has a specific scheme associated with it. So I believe if the selector was updated to include the interactive input scheme it might include that document as part of a notebook. 

Here's the two schemes in use for 'cells' in an interactive window:

```
export const InteractiveInputScheme = 'vscode-interactive-input';
export const InteractiveScheme = 'vscode-interactive';
```

The interactive window isn't tied to a specific language (we've just only used it for python so for).

**Comment by debonte** [comment]
@dbaeumer, here's a picture of the Interactive Window in case you're not familiar with it. When the user clicks the triangular "Execute Code" button to the left of the input box, the text in the input box is removed and placed in a new cell at the bottom of the notebook.

![image](https://user-images.githubusercontent.com/12820898/173112288-6c4d0bac-3294-4437-89ac-5abd80b6ae94.png)

**Comment by rchiodo** [comment]
> May I ask how you get that cell with a different schema into the editor in the first place. I was always under the assumption that VS Code manages all these cells?

VS code does manage the entire interactive window. None of this is created in the extension. It's the one that knows about the schema. We can't change the language as we don't create it.

**Comment by debonte** [comment]
We're still discussing how to properly fix this, but a temporary workaround will be available in our next release.

**Comment by debonte** [comment]
A more correct fix has been checked in and in coming days can be tested with:

- Jupyter 2022.7.1002051049 or later
- Python <tonight's build> or later
- Pylance 2022.7.51 or later (should be available Wednesday 7/27)

**Comment by heejaechang** [comment]
This issue has been fixed in version 2022.7.42, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2022742-27-July-2022)
